### ISSUE TITLE
Unexpected behavior using Dynamic utilities; `matchUtilities` is generating invalid CSS

### ISSUE THREAD

**Issue by elussich** [tailwindlabs/tailwindcss#8648]
**What version of Tailwind CSS are you using?**
TailwindCSS v3.1.3

**What build tool (or framework if it abstracts the build tool) are you using?**
Eleventy v1.0.1

**What version of Node.js are you using?**
Node v16.14.2

**What browser are you using?**
Chrome v101.0.4951.64 (Official Build) (arm64)

**What operating system are you using?**
macOS Big Sur 11.6.5

**Reproduction URL**
https://stackblitz.com/edit/github-ppbygr-jgjnch

**Describe your issue**
I was trying to create a dynamic utility for the `-webkit-text-fill-color` property, using the `matchUtilities` method described here: https://tailwindcss.com/docs/plugins#dynamic-utilities. However, after building the project, the resulting CSS is not what I'm expecting; as a matter of fact, it is not even valid CSS, since the syntax becomes broken, and the passed values do not seem to be correctly applied.

***Steps to reproduce***
Go to this Stackblitz project, https://stackblitz.com/edit/github-ppbygr-jgjnch, do `npm run build` at the console, and then go to the resulting CSS at `./_site/assets/main.css`. At the end of that file you should be able to see the produced CSS from the applied TailwindCSS plugins.

For demonstration purposes I also included another dynamic utility based on tab sizes, which seem to produce the intended result.

The markup file that produces these classes is located at `./src/pages/work.njk`, line 10.

I'm attaching some screenshots for illustration purposes as well here:
<img width="801" alt="tw-issue-markup" src="https://user-images.githubusercontent.com/110324/173882142-1baaa48c-fb18-48c0-8172-98c70865e73c.png">
<img width="803" alt="tw-issue-config" src="https://user-images.githubusercontent.com/110324/173882156-c5975094-9c68-4043-ac2d-87a80704e710.png">
<img width="799" alt="tw-issue-resulting-CSS" src="https://user-images.githubusercontent.com/110324/173882161-a5e3cd5d-2a60-4e3a-bf7d-1b03ee23c02b.png">

Let me know of anything else you may need to throw some light on this issue.
Thank you very much! TailwindCSS is a gas.

**Comment by adamwathan** [comment]
Hey! So in this situation it's because the `theme('colors')` object is this big nested object and hasn't been flattened. We need to come up with a better user-land API for this but you can import our `flattenColorPalette` helper to deal with this in the mean time:

https://play.tailwindcss.com/MByFytk5Pk?file=config

```js
const plugin = require('tailwindcss/plugin');
const flattenColorPalette = require('tailwindcss/lib/util/flattenColorPalette')

module.exports = {
  plugins: [
    plugin(function ({ matchUtilities, theme }) {
      matchUtilities(
        {
          'text-fill': (value) => ({
            '-webkit-text-fill-color': value,
          }),
        },
        { values: flattenColorPalette(theme('colors')) }
      );
    }),
  ],
};
```

Basically it takes an object like this:

```js
let colors = {
  red: {
    500: '...',
    600: '...',
    700: '...',
  },
  blue: {
    500: '...',
    600: '...',
    700: '...',
  }
}
```

...and turns it into this:

```js
let colors = {
  "red-500": "...",
  "red-600": "...",
  "red-700": "...",
  "blue-500": "...",
  "blue-600": "...",
  "blue-700": "...",
}
```

Going to close as not a bug per se that we actually plan to "fix" but definitely want to make this easier without pulling in that extra internal function 👍🏻

**Comment by elussich** [comment]
Thank you very much! This totally works.

I updated Stackblitz with the proposed solution: https://stackblitz.com/edit/github-ppbygr-stjhxn.

**Just one small note**; for some reason, in my example I had to use the `default` property within `flattenColorPalette` package to make it work:
```js
flattenColorPalette.default(theme('colors'))
```

Look forward to a more terse solution, but for now this is just what I need.
Thanks again!
